### PR TITLE
Handle vehicles with multiple engines

### LINKED ISSUE #17298: Some vehicles still affected by 'this vehicle doesn't have an engine' bug
- Solar Car
- Electric Sports Car
- Atomic Sports Car
- Atomic Mini-Tank (tanks mod)

### PR DESCRIPTION
**Pull request by mugling**
Fixes #17298

Our current implementation of multiple engines is somewhat complex. Given that at least one engine must be enabled at all times this fix follows the principle of least surprise:
- When a new engine is installed it remains off by default
- For spawned vehicles (which have no engines enabled) activate all engines when first started
- For constructed vehicles do the same the first time they are started
- Don't otherwise change active engines when starting vehicles

### CODE REVIEW

**Comment by Coolthulhu** [comment]
That `none_of` in `if` looks horrible. I did something like this before and I have no good idea on how to make it look not horrible, but it would be cool if we had a standard for it.
Maybe extracting the lambda instead of having it inlined?

**Comment by codemime** [comment]
Something like that IMO looks fine (and makes the comment redundant):
```c++
const bool no_enabled_engines = std::none_of( engines.begin(), engines.end(), [&]( int idx ) {
    return parts[ idx ].hp > 0 && parts[ idx ].enabled;
};

if( no_enabled_engines ) {
    ...
```

**Comment by mugling** [comment]
Ok how about that

File: src/vehicle.cpp [hunk 1]
```diff
@@ -1460,13 +1460,14 @@ bool vehicle::start_engine( const int e )
 
 void vehicle::start_engines( const bool take_control )
 {
-    int has_engine = false;
     int start_time = 0;
 
-    // if no engines enabled then enable all before trying to start the vehicle
-    if( std::none_of( engines.begin(), engines.end(), [&]( int idx ) {
+    int has_engine = std::count_if( engines.begin(), engines.end(), [&]( int idx ) {
```
Comment (codemime) [inline src/vehicle.cpp]: The name suggests `bool`. I'd replace it with `engine_count` or `num_of_engines`.

**Comment by mugling** [comment]
Or even better that

File: src/vehicle.cpp [hunk 2]
```diff
@@ -1460,16 +1460,22 @@ bool vehicle::start_engine( const int e )
...
+            if( parts[ idx ].hp > 0 ) {
+                parts[ idx ].enabled = true;
+            }
+        }
     }
 
+    int start_time = 0;
     for( size_t e = 0; e < engines.size(); ++e ) {
-        has_engine = has_engine || is_engine_on( e );
+        has_engine += is_engine_on( e );
```
Comment (codemime) [inline src/vehicle.cpp]: Since it's `bool`, `has_engine |= ...` is cleaner.

Comment (Coolthulhu) [inline src/vehicle.cpp]: `int` operations on `bool`s are always ugly.

I'd expect it to wrap around.

Comment (mugling) [inline src/vehicle.cpp]: > I'd expect it to wrap around.

Only after a **lot** of engines but point taken

**Pull request merged**